README.md
# GrowthForecast Client [![Build Status](https://secure.travis-ci.org/sonots/growthforecast-client.png?branch=master)](http://travis-ci.org/sonots/growthforecast-client)

testing ruby: 1.9.2, 1.9.3, 2.0.0; GrowthForecast: >= 0.62 (Jun 27, 2013 released)

## About GrowthForecast Client

growthforecast-client is a ruby client library for GrowthForecast API where [GrowthForecast](http://kazeburo.github.com/GrowthForecast/) is a visualization graph tool.

With growthforecast-client, for example, you can edit properties of a graph such as `color`, and create a complex graph.

## USAGE

    gem install growthforecast-client

See [examples](examples) directory.

### CLI

GrowthForecast Client also provides a CLI named `growthforecast-client`.

Delete a graph or graphs under a path (copy and paste the URL from your GrowthForecast): 

```
$ growthforecast-client delete 'http://fqdn.to.growthforecast:5125/list/service_name'
```

Change the colors of graphs

```
$ growthforecast-client color 'http://fqdn.to.growthforecast:5125/list/service_name' -c 'graph_name:#1111cc' ...
```

See help for more:

```
$ growthforecast-client help
```

### Tips

#### Debug Print

Following codes prints the http requests and responses to STDOUT

```
client = GrowthForecast::Client.new('http://localhost:5125')
client.debug_dev = STDOUT # IO object
```

## Contributing

1. Fork it
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create new [Pull Request](../../pull/new/master)

## Copyright

Copyright (c) 2013 Naotoshi Seo. See [LICENSE](LICENSE) for details.


CHANGELOG.md
# 0.82.5 (2014/06/12)

Fixes:

  - Fix for uninitialized constant OpenSSL (thanks @kui)

# 0.82.4 (2014/06/10)

Fixes:

  - Rescue exception on `bench` command

# 0.82.3 (2014/06/10)

Enhancement:

  - Add `bench` command

# 0.82.2 (2014/03/26)

Enhancement:

  - Add `vrule` command

# 0.82.1 (2014/03/26)

Changes:

  - Changed the argument order of #post_vrule

# 0.82.0 (2014/03/26)

Enhancement:

  - Support newly created `vrule` api

# 0.80.2 (2014/03/06)

Fixes:

  - Fixed handle_error

# 0.80.1 (2014/02/11)

Fixes:

  - Fixed `net/http` errors

# 0.80.0 (2014/02/03)

Changes:

  - Use `net/http` instead of `httpclient` because we met difficulties in multi-thread environments

# 0.62.4 (2014/02/01)

Features:

  - Add `keepalive` option
  - Proxy httpclient parameters

# 0.62.3 (2013/09/20)

Features:

  - Add -g and -s options to `delete` command

# 0.62.2 (2013/09/19)

Changes:

  - Change option -u to an argument of `color`, `create_complex` command

# 0.62.1 (2013/09/19)

Features:

  - Add `growthforecast-client color` command
  - Add `growthforecast-client create_complex` command

Changes:

  - Change `growthforecast-client delete_graph` to `delete` command

# 0.62.0 (2013/07/02)

Changes:

  - Version up!

# 0.0.6 (2013/05/17)

Features:

  - Add `#base_uri` method

# 0.0.5 (2013/05/16)

Features:

  - Add `#last_response` method to get HTTPClient response directory
  - Add `#client=` method to replace HTTPClient object

# 0.0.3
Add bin/growthforecast-client delete_graph


spec/spec_helper.rb
# encoding: utf-8
require "bundler/setup"

ENV['MOCK'] ||= 'on'
require "pry"
require 'growthforecast-client'
require 'webmock/rspec'
require 'cgi'
WebMock.allow_net_connect! if ENV['MOCK'] == 'off'

ROOT = File.dirname(__FILE__)
Dir[File.expand_path("support/**/*.rb", ROOT)].each {|f| require f }

RSpec.configure do |config|
  config.treat_symbols_as_metadata_keys_with_true_values = true
  config.run_all_when_everything_filtered = true
end

def u(str)
  ::CGI.unescape(str.gsub('%20', '+')) if str
end

def e(str)
  ::CGI.escape(str).gsub('+', '%20') if str
end

def base_uri
  'http://localhost:5125'
end

def client
  GrowthForecast::Client.new(base_uri)
end



spec/support/mock.rb
# -*- encoding: utf-8 -*-

shared_context "stub_list_graph" do
  def list_graph_example
    [
      {"service_name"=>"app name",
       "section_name"=>"host name",
       "graph_name"=>"<1sec count",
       "id"=>1},
      {"service_name"=>"app name",
       "section_name"=>"host name",
       "graph_name"=>"<2sec count",
       "id"=>2},
    ]
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/json/list/graph").to_return(:status => 200, :body => list_graph_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_get_graph" do
  def graph_example
    {
      "number"=>0,
      "llimit"=>-1000000000,
      "mode"=>"gauge",
      "stype"=>"AREA",
      "adjustval"=>"1",
      "meta"=>"",
      "service_name"=>"app name",
      "gmode"=>"gauge",
      "color"=>"#cc6633",
      "created_at"=>"2013/02/02 00:41:11",
      "section_name"=>"host name",
      "ulimit"=>1000000000,
      "id"=>1,
      "graph_name"=>"<1sec count",
      "description"=>"",
      "sulimit"=>100000,
      "unit"=>"",
      "sort"=>0,
      "updated_at"=>"2013/02/02 02:32:10",
      "adjust"=>"*",
      "type"=>"AREA",
      "sllimit"=>-100000,
      "md5"=>"3c59dc048e8850243be8079a5c74d079"
    }
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/api/#{e graph['service_name']}/#{e graph['section_name']}/#{e graph['graph_name']}").
    to_return(:status => 200, :body => graph_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_get_graph_by_id" do
  # /json/graph/:id does not return `meta` and `md5`
  def graph_example
    {
      "number"=>0,
      "llimit"=>-1000000000,
      "mode"=>"gauge",
      "stype"=>"AREA",
      "adjustval"=>"1",
      # "meta"=>"",
      "service_name"=>"app name",
      "gmode"=>"gauge",
      "color"=>"#cc6633",
      "created_at"=>"2013/02/02 00:41:11",
      "section_name"=>"host name",
      "ulimit"=>1000000000,
      "id"=>1,
      "graph_name"=>"<1sec count",
      "description"=>"",
      "sulimit"=>100000,
      "unit"=>"",
      "sort"=>0,
      "updated_at"=>"2013/02/02 02:32:10",
      "adjust"=>"*",
      "type"=>"AREA",
      "sllimit"=>-100000,
      # "md5"=>"3c59dc048e8850243be8079a5c74d079"
    }
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/json/graph/#{graph['id']}").
    to_return(:status => 200, :body => graph_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_post_graph" do
  include_context "stub_get_graph"
  before do
    stub_request(:post, "#{base_uri}/api/#{e graph['service_name']}/#{e graph['section_name']}/#{e graph['graph_name']}").
    to_return(:status => 200, :body => { "error" => 0, "data" => graph_example }.to_json)
  end
end

shared_context "stub_delete_graph" do
  proc = Proc.new do
    stub_request(:post, "#{base_uri}/delete/#{e graph['service_name']}/#{e graph['section_name']}/#{e graph['graph_name']}").
    to_return(:status => 200, :body => { "error" => 0 }.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_delete_graph_by_id" do
  proc = Proc.new do
    stub_request(:post, "#{base_uri}/json/delete/graph/#{id}").
    to_return(:status => 200, :body => { "error" => 0 }.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_edit_graph" do
  include_context "stub_get_graph"

  proc = Proc.new do
    stub_request(:post, "#{base_uri}/json/edit/graph/#{graph['id']}").
    to_return(:status => 200, :body => { "error" => 0 }.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_list_complex" do
  def list_complex_example
    [
      {"service_name"=>"app name",
       "section_name"=>"host name",
       "graph_name"=>"complex graph test",
       "id"=>1},
    ]
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/json/list/complex").
    to_return(:status => 200, :body => list_complex_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_get_complex" do
  def complex_example
    {"number"=>0,
     "complex"=>true,
     "created_at"=>"2013/05/20 15:08:28",
     "service_name"=>"app name",
     "section_name"=>"host name",
     "id"=>1,
     "graph_name"=>"complex graph test",
     "data"=>
    [{"gmode"=>"gauge", "stack"=>false, "type"=>"AREA", "graph_id"=>218},
     {"gmode"=>"gauge", "stack"=>true, "type"=>"AREA", "graph_id"=>217}],
    "sumup"=>false,
    "description"=>"complex graph test",
    "sort"=>10,
    "updated_at"=>"2013/05/20 15:08:28"}
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/json/complex/#{e to_complex['service_name']}/#{e to_complex['section_name']}/#{e to_complex['graph_name']}").
    to_return(:status => 200, :body => complex_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_get_complex_by_id" do
  def complex_example
    {"number"=>0,
     "complex"=>true,
     "created_at"=>"2013/05/20 15:08:28",
     "service_name"=>"app name",
     "section_name"=>"host name",
     "id"=>1,
     "graph_name"=>"complex graph test",
     "data"=>
    [{"gmode"=>"gauge", "stack"=>false, "type"=>"AREA", "graph_id"=>218},
     {"gmode"=>"gauge", "stack"=>true, "type"=>"AREA", "graph_id"=>217}],
    "sumup"=>false,
    "description"=>"complex graph test",
    "sort"=>10,
    "updated_at"=>"2013/05/20 15:08:28"}
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/json/complex/#{id}").
    to_return(:status => 200, :body => complex_example.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_delete_complex" do
  proc = Proc.new do
    stub_request(:post, "#{base_uri}/json/delete/complex/#{e to_complex['service_name']}/#{e to_complex['section_name']}/#{e to_complex['graph_name']}").
    to_return(:status => 200, :body => { "error" => 0 }.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_delete_complex_by_id" do
  proc = Proc.new do
    stub_request(:post, "#{base_uri}/delete_complex/#{id}").
    to_return(:status => 200, :body => { "error" => 0 }.to_json)
  end
  before(:each, &proc)
end

shared_context "stub_create_complex" do
  include_context "stub_list_complex"

  proc = Proc.new do
    list_graph_example.each do |graph|
      stub_request(:get, "#{base_uri}/api/#{e graph['service_name']}/#{e graph['section_name']}/#{e graph['graph_name']}").
      to_return(:status => 200, :body => graph.to_json)
    end

    stub_request(:post, "#{base_uri}/json/create/complex").
    to_return(:status => 200, :body => { "error" => 0 }.to_json)
  end
  before(:each, &proc)
end

def notfound_graph
  {
    'service_name' => 'not found',
    'section_name' => 'not found',
    'graph_name'   => 'not found',
  }
end
shared_context "stub_get_notfound_graph" do
  proc = Proc.new do
    stub_request(:get, "#{base_uri}/api/#{e notfound_graph['service_name']}/#{e notfound_graph['section_name']}/#{e notfound_graph['graph_name']}").
    to_return(:status => 404, :body => '<html><body><strong>404</strong> Not Found</body></html>')
  end
  before(:each, &proc)
end

shared_context "stub_post_vrule" do
  include_context "stub_get_vrule"
  before do
    stub_request(:post, "#{base_uri}/vrule/api/#{e graph['service_name']}/#{e graph['section_name']}/#{e graph['graph_name']}").
    to_return(:status => 200, :body => { "error" => 0, "data" => vrule_example }.to_json)
  end
end

shared_context "stub_get_vrule" do
  def vrule_example
    {
      "graph_path"=>"/#{graph['service_name']}/#{graph['section_name']}/#{graph['graph_name']}",
      "color"=>"#FF0000",
      "time"=>1395826210,
      "id"=>1,
      "dashes"=>"2,10",
      "description"=>""
    }
  end

  proc = Proc.new do
    stub_request(:get, "#{base_uri}/vrule/summary/#{e graph['service_name']}/#{e graph['section_name']}/#{e graph['graph_name']}").
    to_return(:status => 200, :body => [vrule_example].to_json)
  end
  before(:each, &proc)
end

shared_context "stub_get_notfound_vrule" do
  proc = Proc.new do
    stub_request(:get, "#{base_uri}/vrule/summary/#{e notfound_graph['service_name']}/#{e notfound_graph['section_name']}/#{e notfound_graph['graph_name']}").
    to_return(:status => 200, :body => [].to_json)
  end
  before(:each, &proc)
end



spec/support/setup.rb
# -*- encoding: utf-8 -*-

shared_context "setup_growthforecast_client" do
  include_context "stub_list_graph" if ENV['MOCK'] == 'on'
  let(:graphs) { client.list_graph }
  let(:graph) { graphs.first }
  let(:from_graphs) do
    [
      {
        "service_name" => graphs[0]["service_name"],
        "section_name" => graphs[0]["section_name"],
        "graph_name"   => graphs[0]["graph_name"],
        "gmode" => "gauge",
        "stack" => false,
        "type" => "AREA",
      },
      {
        "service_name" => graphs[1]["service_name"],
        "section_name" => graphs[1]["section_name"],
        "graph_name"   => graphs[1]["graph_name"],
        "gmode" => "gauge",
        "stack" => false,
        "type" => "AREA"
      },
    ]
  end
  let(:to_complex) do
    {
      "service_name" => graphs.first["service_name"],
      "section_name" => graphs.first["section_name"],
      "graph_name"   => "complex graph test",
      "description"  => "complex graph test",
      "sort"         => 10
    }
  end

  include_context "stub_post_graph" if ENV['MOCK'] == 'on'
  include_context "stub_delete_graph" if ENV['MOCK'] == 'on'
  unless ENV['MOCK'] == 'on'
    before {
      client.delete_graph("app name", "host name", "<1sec count") rescue nil
      client.delete_graph("app name", "host name", "<2sec count") rescue nil
      client.post_graph("app name", "host name", "<1sec count", { 'number' => 0 }) rescue nil
      client.post_graph("app name", "host name", "<2sec count", { 'number' => 0 }) rescue nil
    }
    after {
      client.delete_graph("app name", "host name", "<1sec count") rescue nil
      client.delete_graph("app name", "host name", "<2sec count") rescue nil
    }
  end
end


examples/edit_graph.rb
# -*- encoding: utf-8 -*-
# An example to edit properties of graphs such as color, unit. 

# require anyway
require 'growthforecast-client'

# Create a GrowthForecast Client, given he base URI of GrowthForecast
client = GrowthForecast::Client.new('http://localhost:5125')
client.debug_dev = STDOUT # debug print the http requests and responses

# configure colors of graphs whose names are as belows:
graph_colors = {
  '<1sec_count'  => '#1111cc',
  '<2sec_count'  => '#11cc11',
  '<3sec_count'  => '#cc7711',
  '<4sec_count'  => '#cccc11',
  '>=4sec_count' => '#cc1111',
}
# I gonna apply for all services/sections
sections = client.list_section
sections.each do |service_name, sections|
  sections.each do |section_name|
    graph_colors.keys.each do |graph_name|
      # Graph properties to overwrite
      params = {
        'color'  => graph_colors[graph_name],
        'unit'   => 'count',
        'sort'   => 1, # order to display, 19 is the top
        'adjust' => '/',
        'adjustval' => '1',
      }
      # Edit a graph
      begin
        puts "Setup /#{service_name}/#{section_name}/#{graph_name}"
        client.edit_graph(service_name, section_name, graph_name, params)
      rescue GrowthForecast::NotFound => e
        puts "\tclass:#{e.class}\t#{e.message}"
      end
    end
  end
end


examples/complex_graph.rb
# -*- encoding: utf-8 -*-
require 'growthforecast-client'

# Create a GrowthForecast Client, given he base URI of GrowthForecast
uri = 'http://localhost:5125'
client = GrowthForecast::Client.new(uri)

# Apply for all services/sections
sections = client.list_section
sections.each do |service_name, sections|
  sections.each do |section_name|
    # Make a complex graph from these graphs
    from_graphs= [
      {:path => "#{service_name}/#{section_name}/<1sec_count", :gmode => 'gauge', :stack => true, :type => 'AREA'},
      {:path => "#{service_name}/#{section_name}/<2sec_count", :gmode => 'gauge', :stack => true, :type => 'AREA'},
      {:path => "#{service_name}/#{section_name}/<3sec_count", :gmode => 'gauge', :stack => true, :type => 'AREA'},
      {:path => "#{service_name}/#{section_name}/<4sec_count", :gmode => 'gauge', :stack => true, :type => 'AREA'},
      {:path => "#{service_name}/#{section_name}/>=4sec_count", :gmode => 'gauge', :stack => true, :type => 'AREA'},
    ]

    # The propety of a complex graph to create, e.g., path
    to_complex = {
      :path         => "#{service_name}/#{section_name}/response_count",
      :description  => 'response time count',
      :sort         => 10,
    }

    begin
      puts "Setup #{to_complex[:path]}"
      client.create_complex(from_graphs, to_complex)
    rescue GrowthForecast::AlreadyExists => e
      puts "\tclass:#{e.class}\t#{e.message}"
    rescue GrowthForecast::NotFound => e
      puts "\tclass:#{e.class}\t#{e.message}"
    end
  end
end


lib/growthforecast/client.rb
# -*- encoding: utf-8 -*-
require 'net/http'
require 'net/https'
require 'uri'
require 'json'
require 'cgi'
require 'openssl'

module GrowthForecast
  class Error < StandardError; end
  class NotFound < Error; end
  class AlreadyExists < Error; end

  class Client
    attr_reader   :base_uri
    attr_reader   :host
    attr_reader   :port
    attr_accessor :debug_dev
    attr_accessor :open_timeout
    attr_accessor :read_timeout
    attr_accessor :verify_ssl
    attr_accessor :keepalive

    # @param [String] base_uri The base uri of GrowthForecast
    def initialize(base_uri = 'http://127.0.0.1:5125', opts = {})
      @base_uri = base_uri
      opts = stringify_keys(opts)

      URI.parse(base_uri).tap {|uri|
        @host       = uri.host
        @port       = uri.port
        @use_ssl    = uri.scheme == 'https'
      }
      @debug_dev    = opts['debug_dev'] # IO object such as STDOUT
      @open_timeout = opts['open_timeout'] # 60
      @read_timeout = opts['read_timeout'] # 60
      @keepalive    = opts['keepalive']
      @verify_ssl   = opts['verify_ssl']
      @ca_file      = opts['ca_file']
    end

    def http_connection
      Net::HTTP.new(@host, @port).tap {|http|
        http.use_ssl      = @use_ssl
        http.open_timeout = @open_timeout if @open_timeout
        http.read_timeout = @read_timeout if @read_timeout
        if @verify_ssl
          http.verify_mode = OpenSSL::SSL::VERIFY_PEER
          http.ca_file     = @ca_file if @ca_file
        else
          http.verify_mode = OpenSSL::SSL::VERIFY_NONE
        end
        http.set_debug_output(@debug_dev) if @debug_dev
      }
    end

    def get_request(path, extheader = {})
      Net::HTTP::Get.new(path).tap {|req|
        req['Host'] = @host
        req['Connection'] = 'Keep-Alive' if @keepalive
        extheader.each {|key, value| req[key] = value }
      }
    end

    def post_request(path, body, extheader = {})
      Net::HTTP::Post.new(path).tap {|req|
        req['Host'] = @host
        req['Connection'] = 'Keep-Alive' if @keepalive
        extheader.each {|key, value| req[key] = value }
        req.body = body
      }
    end

    def last_request_uri
      @request_uri
    end

    def last_response
      @res
    end

    # GET the JSON API
    # @param [String] path
    # @return [Hash] response body
    def get_json(path)
      @request_uri = "#{@base_uri}#{path}" 
      req  = get_request(path)
      @res = http_connection.start {|http| http.request(req) }
      handle_error(@res, @request_uri)
      JSON.parse(@res.body)
    end

    # POST the JSON API
    # @param [String] path
    # @param [Hash] data 
    # @return [Hash] response body
    def post_json(path, data = {})
      @request_uri = "#{@base_uri}#{path}" 
      body = JSON.generate(data)
      extheader = { 'Content-Type' => 'application/json' }
      req  = post_request(path, body, extheader)
      @res = http_connection.start {|http| http.request(req) }
      handle_error(@res, @request_uri)
      JSON.parse(@res.body)
    end

    # POST the non-JSON API
    # @param [String] path
    # @param [Hash] data 
    # @return [String] response body
    def post_query(path, data = {})
      @request_uri = "#{@base_uri}#{path}" 
      body = URI.encode_www_form(data)
      extheader = { 'Content-Type' => 'application/x-www-form-urlencoded' }
      req  = post_request(path, body, extheader)
      @res = http_connection.start {|http| http.request(req) }
      handle_error(@res, @request_uri)
      JSON.parse(@res.body)
    end

    # Get the list of graphs, /json/list/graph
    # @return [Hash] list of graphs
    # @example
    # [
    #   {"service_name"=>"test",
    #    "graph_name"=>"<2sec_count",
    #    "section_name"=>"hostname",
    #    "id"=>4},
    #   {"service_name"=>"test",
    #    "graph_name"=>"<1sec_count",
    #    "section_name"=>"hostname",
    #    "id"=>3},
    # ]
    def list_graph(service_name = nil, section_name = nil, graph_name = nil)
      graphs = get_json('/json/list/graph')
      graphs = graphs.select {|g| g['service_name'] == service_name } if service_name
      graphs = graphs.select {|g| g['section_name'] == section_name } if section_name
      graphs = graphs.select {|g| g['graph_name']   == graph_name   } if graph_name
      graphs
    end

    # A Helper: Get the list of section
    # @return [Hash] list of sections
    # @example
    # {
    #   "service_name1" => [
    #     "section_name1",
    #     "section_name2",
    #   ],
    #   "service_name2" => [
    #     "section_name1",
    #     "section_name2",
    #   ],
    # }
    def list_section(service_name = nil, section_name = nil, graph_name = nil)
      graphs = list_graph(service_name, section_name, graph_name)
      services = {}
      graphs.each do |graph|
        service_name, section_name = graph['service_name'], graph['section_name']
        services[service_name] ||= {}
        services[service_name][section_name] ||= true
      end
      Hash[services.map {|service_name, sections| [service_name, sections.keys] }]
    end

    # A Helper: Get the list of services
    # @return [Array] list of services
    # @example
    # [
    #   "service_name1",
    #   "service_name2",
    # ]
    def list_service(service_name = nil, section_name = nil, graph_name = nil)
      graphs = list_graph(service_name, section_name, graph_name)
      services = {}
      graphs.each do |graph|
        service_name = graph['service_name']
        services[service_name] ||= true
      end
      services.keys
    end

    # Post parameters to a graph, POST /api/:service_name/:section_name/:graph_name
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @param [Hash] params The POST parameters. See #get_graph
    # @return [Hash] the error code and graph property
    # @example
    #{"error"=>0,
    #"data"=>{
    #  "number"=>1,
    #  "llimit"=>-1000000000,
    #  "mode"=>"gauge",
    #  "stype"=>"AREA",
    #  "adjustval"=>"1",
    #  "meta"=>"",
    #  "service_name"=>"test",
    #  "gmode"=>"gauge",
    #  "color"=>"#cc6633",
    #  "created_at"=>"2013/02/02 00:41:11",
    #  "section_name"=>"hostname",
    #  "ulimit"=>1000000000,
    #  "id"=>21,
    #  "graph_name"=>"<4sec_count",
    #  "description"=>"",
    #  "sulimit"=>100000,
    #  "unit"=>"",
    #  "sort"=>0,
    #  "updated_at"=>"2013/02/02 02:32:10",
    #  "adjust"=>"*",
    #  "type"=>"AREA",
    #  "sllimit"=>-100000,
    #  "md5"=>"3c59dc048e8850243be8079a5c74d079"}}
    def post_graph(service_name, section_name, graph_name, params)
      post_query("/api/#{e service_name}/#{e section_name}/#{e graph_name}", params)
    end

    # Get the propety of a graph, GET /api/:service_name/:section_name/:graph_name
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @return [Hash] the graph property
    # @example
    #{
    #  "number"=>1,
    #  "llimit"=>-1000000000,
    #  "mode"=>"gauge",
    #  "stype"=>"AREA",
    #  "adjustval"=>"1",
    #  "meta"=>"",
    #  "service_name"=>"test",
    #  "gmode"=>"gauge",
    #  "color"=>"#cc6633",
    #  "created_at"=>"2013/02/02 00:41:11",
    #  "section_name"=>"hostname",
    #  "ulimit"=>1000000000,
    #  "id"=>21,
    #  "graph_name"=>"<4sec_count",
    #  "description"=>"",
    #  "sulimit"=>100000,
    #  "unit"=>"",
    #  "sort"=>0,
    #  "updated_at"=>"2013/02/02 02:32:10",
    #  "adjust"=>"*",
    #  "type"=>"AREA",
    #  "sllimit"=>-100000,
    #  "md5"=>"3c59dc048e8850243be8079a5c74d079"}
    def get_graph(service_name, section_name, graph_name)
      get_json("/api/#{e service_name}/#{e section_name}/#{e graph_name}")
    end

    # Get the propety of a graph, /json/graph/:id
    # @param [String] id
    # @return [Hash] the graph property
    # @example
    # {"llimit"=>-1000000000,
    #  "number"=>48778224,
    #  "stype"=>"AREA",
    #  "mode"=>"count",
    #  "complex"=>false,
    #  "adjustval"=>"1",
    #  "created_at"=>"2013/02/01 16:01:17",
    #  "color"=>"#cc3366",
    #  "service_name"=>"test",
    #  "gmode"=>"gauge",
    #  "ulimit"=>1000000000,
    #  "section_name"=>"all",
    #  "id"=>1,
    #  "graph_name"=>"response_time_max",
    #  "description"=>"",
    #  "sort"=>0,
    #  "unit"=>"",
    #  "sulimit"=>100000,
    #  "updated_at"=>"2013/02/04 18:26:49",
    #  "adjust"=>"*",
    #  "sllimit"=>-100000,
    #  "type"=>"AREA"}
    def get_graph_by_id(id)
      get_json("/json/graph/#{id}")
    end

    # Delete a graph, POST /delete/:service_name/:section_name/:graph_name
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    def delete_graph(service_name, section_name, graph_name)
      post_query("/delete/#{e service_name}/#{e section_name}/#{e graph_name}")
    end

    # Delete a graph, POST /json/delete/graph/:id
    # @param [String] id
    def delete_graph_by_id(id)
      post_query("/json/delete/graph/#{id}")
    end

    # Update the property of a graph, /json/edit/graph/:id
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @param [Hash]   params
    #   All of parameters given by #get_graph are available except `number` and `mode`.
    # @return [Hash]  error response
    # @example
    # {"error"=>0} #=> Success
    # {"error"=>1} #=> Error
    def edit_graph(service_name, section_name, graph_name, params)
      data = get_graph(service_name, section_name, graph_name)
      id = data['id']
      updates = handle_update_params(data, params)
      post_json("/json/edit/graph/#{id}", updates)
    end

    # Get the list of complex graphs, /json/list/complex
    # @return [Hash] list of complex graphs
    # @example
    # [
    #   {"service_name"=>"test",
    #    "graph_name"=>"<2sec_count",
    #    "section_name"=>"hostname",
    #    "id"=>4},
    #   {"service_name"=>"test",
    #    "graph_name"=>"<1sec_count",
    #    "section_name"=>"hostname",
    #    "id"=>3},
    # ]
    def list_complex(service_name = nil, section_name = nil, graph_name = nil)
      graphs = get_json('/json/list/complex')
      graphs = graphs.select {|g| g['service_name'] == service_name } if service_name
      graphs = graphs.select {|g| g['section_name'] == section_name } if section_name
      graphs = graphs.select {|g| g['graph_name']   == graph_name   } if graph_name
      graphs
    end

    # Create a complex graph
    #
    # @param [Array] from_graphs Array of graph properties whose keys are
    #   ["service_name", "section_name", "graph_name", "gmode", "stack", "type"]
    # @param [Hash] to_complex Property of Complex Graph, whose keys are like
    #   ["service_name", "section_name", "graph_name", "description", "sort"]
    def create_complex(from_graphs, to_complex)
      graph_data = []
      from_graphs.each do |from_graph|
        graph = get_graph(from_graph["service_name"], from_graph["section_name"], from_graph["graph_name"])
        graph_id = graph['id']

        graph_data << {
          :gmode    => from_graph["gmode"],
          :stack    => from_graph["stack"],
          :type     => from_graph["type"],
          :graph_id => graph_id
        }
      end

      post_params = {
        :service_name => to_complex["service_name"],
        :section_name => to_complex["section_name"],
        :graph_name   => to_complex["graph_name"],
        :description  => to_complex["description"],
        :sort         => to_complex["sort"],
        :data         => graph_data
      }
      post_json('/json/create/complex', post_params)
    end

    # Get the propety of a complex graph, GET /json/complex/:service_name/:section_name/:graph_name
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @return [Hash] the graph property
    # @version 0.70 or more
    # @example
    # {"number"=>0,
    #  "complex"=>true,
    #  "created_at"=>"2013/05/20 15:08:28",
    #  "service_name"=>"app name",
    #  "section_name"=>"host name",
    #  "id"=>18,
    #  "graph_name"=>"complex graph test",
    #  "data"=>
    #   [{"gmode"=>"gauge", "stack"=>false, "type"=>"AREA", "graph_id"=>218},
    #    {"gmode"=>"gauge", "stack"=>true, "type"=>"AREA", "graph_id"=>217}],
    #  "sumup"=>false,
    #  "description"=>"complex graph test",
    #  "sort"=>10,
    #  "updated_at"=>"2013/05/20 15:08:28"}
    def get_complex(service_name, section_name, graph_name)
      get_json("/json/complex/#{e service_name}/#{e section_name}/#{e graph_name}")
    end

    # Get the propety of a complex graph, GET /json/complex/:id
    # @param [String] id
    # @return [Hash] the graph property
    # @version 0.70 or more
    # @example See #get_complex
    def get_complex_by_id(id)
      get_json("/json/complex/#{id}")
    end

    # Delete a complex graph
    #
    # This is a helper method of GrowthForecast API to find complex graph id and call delete_complex_by_id
    #
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @return [Hash]  error response
    # @example
    # {"error"=>0} #=> Success
    # {"error"=>1} #=> Error
    def delete_complex(service_name, section_name, graph_name)
      post_query("/json/delete/complex/#{e service_name}/#{e section_name}/#{e graph_name}")
    end

    # Delete a complex graph, /delete_complex/:complex_id
    #
    # @param [String] id of a complex graph
    # @return [Hash]  error response
    # @example
    # {"error"=>0} #=> Success
    # {"error"=>1} #=> Error
    def delete_complex_by_id(complex_id)
      post_query("/delete_complex/#{complex_id}")
    end

    # Post parameters to a vrule, POST /vrule/api/:service_name/:section_name/:graph_name
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @param [Hash] params The POST parameters. See #get_vrule
    # @return [Hash] the error code and graph property
    # @example
    #{"error"=>0,
    #"data"=>{
    #  "graph_path"=>"/hoge/hoge/hoge",
    #  "color"=>"#FF0000",
    #  "time"=>1395826210,
    #  "id"=>1,
    #  "dashes"=>"2,10",
    #  "description"=>""}}
    def post_vrule(service_name = nil, section_name = nil, graph_name = nil, params = {})
      path = "/vrule/api"
      path += "/#{e service_name}" if service_name
      path += "/#{e section_name}" if section_name
      path += "/#{e graph_name}"   if graph_name
      post_query(path, params)
    end

    # Get the data of vrules, GET /vrule/summary/:service_name/:section_name/:graph_name
    # @param [String] service_name
    # @param [String] section_name
    # @param [String] graph_name
    # @return [Hash] the data of vrules
    # @example
    #[
    #{
    #  "graph_path"=>"/hoge/hoge/hoge",
    #  "color"=>"#FF0000",
    #  "time"=>1395826210,
    #  "id"=>1,
    #  "dashes"=>"",
    #  "description"=>""
    #},
    #{
    #  "graph_path"=>"/hoge/hoge/hoge",
    #  "color"=>"#FF0000",
    #  "time"=>1395826363,
    #  "id"=>2,
    #  "dashes"=>"2,10",
    #  "description"=>""
    #}
    #]
    def get_vrule(service_name = nil, section_name = nil, graph_name = nil)
      path = "/vrule/summary"
      path += "/#{e service_name}" if service_name
      path += "/#{e section_name}" if section_name
      path += "/#{e graph_name}"   if graph_name
      get_json(path)
    end

    private

    def e(str)
      CGI.escape(str).gsub('+', '%20') if str
    end

    def handle_error(res, request_uri)
      case res.code
      when '200'
      when '404'
        raise NotFound.new(error_message(res, request_uri))
      when '409'
        raise AlreadyExists.new(error_message(res, request_uri))
      else
        raise Error.new(error_message(res, request_uri))
      end
    end

    def error_message(res, request_uri)
      "status:#{res.code}\turi:#{request_uri}\tmessage:#{res.body}"
    end

    # GrowthForecast's /json/edit/graph API requires all parameters to update, thus
    # we have to merge the original graph parameters and parameters which we want to update. Sucks!
    #
    # @param [Hash] graph_data the current graph property data
    # @param [Hash[ params     the parameters which we want to update
    # @return [Hash] merged parameters
    def handle_update_params(graph_data, params)
      updates = graph_data.merge(params)
      # `meta` field is automatically added when we call get_graph.
      # If we post `meta` data to update graph, `meta` is constructed circularly. Sucks!
      # Thus, I remove the `meta` here.
      updates['meta'] = '' if !params.has_key?('meta')
      updates
    end

    def stringify_keys(hash)
      {}.tap {|h| hash.each {|key, val| h[key.to_s] = val } }
    end
  end
end

